pkg/readiness/ready_tracker.go

```diff
@@ -689,58 +732,83 @@ func (t *Tracker) trackConstraintTemplates(ctx context.Context) error {
-// Fails-open if the Config resource cannot be fetched or does not exist.
-func (t *Tracker) trackConfig(ctx context.Context) error {
-	var wg sync.WaitGroup
+// trackSyncSources sets expectations for cached data as specified by the singleton Config resource.
+// and any SyncSet resources present on the cluster.
+// Works best effort and fails-open if the a resource cannot be fetched or does not exist.
+func (t *Tracker) trackSyncSources(ctx context.Context) error {
 	defer func() {
-		defer t.config.ExpectationsDone()
-		log.V(1).Info("config expectations populated")
+		t.config.ExpectationsDone()
+		log.V(logging.DebugLevel).Info("config expectations populated")
 
-		wg.Wait()
+		t.syncsets.ExpectationsDone()
+		log.V(logging.DebugLevel).Info("syncset expectations populated")
 	}()
 
+	dataGVKs := make(map[schema.GroupVersionKind]struct{})
+
 	cfg, err := t.getConfigResource(ctx)
 	if err != nil {
-		return fmt.Errorf("fetching config resource: %w", err)
+		log.Error(err, "fetching config resource")
 	}
 	if cfg == nil {
 		log.Info("config resource not found - skipping for readiness")
-		return nil
+	} else {
+		if !cfg.GetDeletionTimestamp().IsZero() {
+			log.Info("config resource is being deleted - skipping for readiness")
+		} else {
+			t.config.Expect(cfg)
+			log.V(logging.DebugLevel).Info("setting expectations for config", "configCount", 1)
+
+			for _, entry := range cfg.Spec.Sync.SyncOnly {
+				dataGVKs[entry.ToGroupVersionKind()] = struct{}{}
+			}
+		}
 	}
-	if !cfg.GetDeletionTimestamp().IsZero() {
-		log.Info("config resource is being deleted - skipping for readiness")
+
+	// Without validation operations, there is no reason to wait for referential data when deciding readiness.
+	if !operations.HasValidationOperations() {
 		return nil
 	}
 
-	if operations.HasValidationOperations() {
-		// Expect the resource kinds specified in the Config.
-		// We will fail-open (resolve expectations) for GVKs
-		// that are unregistered.
-		for _, entry := range cfg.Spec.Sync.SyncOnly {
-			gvk := schema.GroupVersionKind{
-				Group:   entry.Group,
-				Version: entry.Version,
-				Kind:    entry.Kind,
-			}
-			u := &unstructured.Unstructured{}
-			u.SetGroupVersionKind(gvk)
-			t.config.Expect(u)
-			t.config.Observe(u) // we only care about the gvk entry in kinds()
-
-			// Set expectations for individual cached resources
-			dt := t.ForData(gvk)
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				err := t.trackData(ctx, gvk, dt)
-				if err != nil {
-					log.Error(err, "aborted trackData", "gvk", gvk)
+	syncsets := &syncsetv1alpha1.SyncSetList{}
+	lister := retryLister(t.lister, retryAll)
+	if err := lister.List(ctx, syncsets); err != nil {
+		log.Error(err, "listing syncsets")
+	} else {
+		log.V(logging.DebugLevel).Info("setting expectations for syncsets", "syncsetCount", len(syncsets.Items))
+
+		for i := range syncsets.Items {
+			syncset := syncsets.Items[i]
```

Comment:
I'm not sure we should do this since it's being called out as less efficient by the linter:

```
pkg/readiness/ready_tracker.go:780:3: rangeValCopy: each iteration copies 288 bytes (consider pointers or indexing) (gocritic)
                for _, syncset := range syncsets.Items {
                ^
```